src/codegen/compile.ml

```diff
@@ -8626,21 +8193,222 @@ module Stabilization = struct
     | _ -> assert false
 end
 
-module GCRoots = struct
-  let register env static_roots =
+(* New stable memory layout with new version.
+   Prevents forward compatibility of old compiled programs that rely on deserialization.
+  If size == 0: empty
+  If logical size N > 0:
+    [0..4]          0 (first word is backed up at `end-8`)
+    [4..N]          <stable memory>
+            <zero padding>
+    [end-16..end-8] <size N>
+    [end-8..end-4]  <first word>
+    [end-4..end]    <new version>
```

Comment:
I think there's some off-by-1s here:

```suggestion
  let end = physical size * page_size
  If logical size N > 0:
    [0..3]          0 (first word is backed up at `end-8`)
    [4..N-1]          <stable memory>
            <zero padding>
    [end-16..end-9] <size N>
    [end-8..end-5]  <first word>
    [end-4..end-1]    <new version>
```
or maybe:

```suggestion
  let end = physical size * page_size
  If logical size N > 0:
    [0..4)          0 (first word is backed up at `end-8`)
    [4..N)          <stable memory>
            <zero padding>
    [end-16..end-8) <size N>
    [end-8..end-4)  <first word>
    [end-4..end)   <new version>
```

Reply:
Thanks for improvement.